sdk/src/main/java/com/bugsnag/android/BlockedThreadDetector.java

```diff
@@ -0,0 +1,105 @@
+package com.bugsnag.android;
+
+import android.os.Handler;
+import android.os.Looper;
+import android.os.SystemClock;
+
+/**
+ * Detects whether a given thread is blocked by continuously posting a {@link Runnable} to it
+ * from a watcher thread, invoking a delegate if the message is not processed within
+ * a configured interval.
+ */
+final class BlockedThreadDetector {
+
+    static final int MIN_CHECK_INTERVAL_MS = 1000;
+
+    interface Delegate {
+
+        /**
+         * Invoked when a given thread has been unable to execute a {@link Runnable} within
+         * the {@link #blockedThresholdMs}
+         *
+         * @param thread the thread being monitored
+         */
+        void onThreadBlocked(Thread thread);
+    }
+
+    final Looper looper;
+    final long checkIntervalMs;
+    final long blockedThresholdMs;
+    final Handler handler;
+    final Delegate delegate;
+
+    volatile long lastUpdateMs;
+    volatile boolean isAlreadyBlocked = false;
+
+    BlockedThreadDetector(long blockedThresholdMs,
+                          Looper looper,
+                          Delegate delegate) {
+        this(blockedThresholdMs, MIN_CHECK_INTERVAL_MS, looper, delegate);
+    }
+
+    BlockedThreadDetector(long blockedThresholdMs,
+                          long checkIntervalMs,
+                          Looper looper,
+                          Delegate delegate) {
+        if ((blockedThresholdMs <= 0 || checkIntervalMs <= 0
+            || looper == null || delegate == null)) {
+            throw new IllegalArgumentException();
+        }
+        this.blockedThresholdMs = blockedThresholdMs;
+        this.checkIntervalMs = checkIntervalMs;
+        this.looper = looper;
+        this.delegate = delegate;
+        this.handler = new Handler(looper);
+    }
+
+    void updateLivenessTimestamp() {
+        lastUpdateMs = SystemClock.elapsedRealtime();
+    }
+
+    void start() {
+        updateLivenessTimestamp();
+        handler.post(livenessCheck);
+        watcherThread.start();
+    }
+
+    final Runnable livenessCheck = new Runnable() {
+        @Override
+        public void run() {
+            updateLivenessTimestamp();
+            handler.postDelayed(this, checkIntervalMs);
+        }
+    };
+
+    final Thread watcherThread = new Thread() {
+        @Override
+        public void run() {
+            while (!isInterrupted()) {
+                // when we would next consider the app blocked if no timestamp updates take place
+                long now = SystemClock.elapsedRealtime();
```

Comment:
"uptimeMillis" probably works better here, as time in deep sleep shouldn't be accounted for as "application not responding"